### ISSUE TITLE
Null TreePaths in finally blocks

### ISSUE THREAD

**Issue by CharlesZ-Chen** [typetools/checker-framework#1520]
Below code is a minimum test case:

```java
public class C {
    void start() {
        new Runnable() {
            public void run() {
                try {
                    _run();
                } finally {
                    signal(); //Evaluating this node type as member of implicit `this` will throw NPE
                }
            }
        };
    }

    void signal() {
    }

    void _run() {
    }
}
```

Running Checker Framework (version `2.2.0`) on above code with any checker, will cause a Null Pointer Exception thrown from `AnnotatedTypes.asMemberOfImpl()`.

The reason might be `AnnotatedTypeFactory#getPath(tree)` failed to find the tree path of method invocation tree `signal()` --- which cause `ATF#getImplicitReceiverType(tree)` return `null` for `signal()`, and this cause `AnnotatedTypes.asMemberOfImpl()` throws the NPE.

At first I suspected if `treePathCache` didn't cache path correctly in [`AnnotatedTypeFactory#getPath(tree)`](https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java#L2715), and I tried modify the last return statement from `return treePathCache.getPath(root, node);` to `TreePath.get(root, node)` (just as the way before introducing tree path cacheing), and surprisedly found even `TreePath` failed to find the path of node `signal()`, given the `root` is the complete compilation unit (i.e. the complete CompilationUnitTree of `MultipointServer`).

Consider `TreePath` is a class from package `com.sun.source.util`, so is this a javac bug?

Any ideas are highly appreciated!

**Comment by wmdietl** [comment]
Thanks for the report!
Verified the issue with master and can reproduce the crash.

**Comment by CharlesZ-Chen** [comment]
Update: It seems this crash was introduced recently --- running CF 2.1.14 works fine with above code, while CF 2.2.0 crashes on above code.

**Comment by wmdietl** [comment]
@bitterfox Could you have a look at this issue?

**Comment by CharlesZ-Chen** [comment]
Another similar code seems related to this issue was found:

```java
class TopLevelClass {
    protected class InnerClass implements Runnable {
        public void run() {
            try {
                if (true) {
                    throw new RuntimeException();
                }
            } finally {
                topLevelClassMethod();
            }
        }
    }
    void topLevelClassMethod() {
    }
}
```

Running CF `2.2.0` with any Checker on above code will cause CF crashes:

```
Exception: java.lang.NullPointerException; Stack trace: org.checkerframework.framework.util.AnnotatedTypes.asMemberOfImpl(AnnotatedTypes.java:261)
  org.checkerframework.framework.util.AnnotatedTypes.asMemberOf(AnnotatedTypes.java:242)
  org.checkerframework.framework.util.AnnotatedTypes.asMemberOf(AnnotatedTypes.java:204)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1980)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1947)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1469)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:170)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:45)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:54)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1202)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:976)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:202)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:888)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:94)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:79)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:405)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:226)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1155)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1112)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1042)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1418)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:248)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:285)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:164)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:962)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:502)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:185)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1343)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1297)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:902)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:861)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
1 error
```

While CF `2.1.14` works fine with above code.

**Comment by wmdietl** [comment]
This comment probably touches on this same issue: https://github.com/typetools/checker-framework/pull/1346#issuecomment-333604872
@bitterfox maybe for the additional trees for the cloned finally blocks, we need to keep track of the enclosing class or maybe even the whole path?

**Comment by smillst** [comment]
I've added a work around for this particular symptom of the null TreePath problem.  Please revert  https://github.com/typetools/checker-framework/commit/157049f8444dd85562da761be300d79685368a9c when fixing this issue.

**Comment by wmdietl** [comment]
My understanding at the moment: we duplicate finally blocks and some other trees and store them in `org.checkerframework.framework.type.GenericAnnotatedTypeFactory.getGeneratedTrees(Tree)` (coming from the `CFGBuilder` originally).

`BaseTypeVisitor.visitTry` (around https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java#L897 ) visits all the different possible Trees for the finally block.

When we handle a Lambda expression we try to determine the path to it, using `trees.getPath` in `CFGBuilder.CFGTranslationPhaseOne.process` around https://github.com/typetools/checker-framework/blob/master/dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java#L1584.
However, when the Lambda is within one of the duplicate finally Trees, we can't find the TreePath for it.

I think we need to do something like this:
- move `org.checkerframework.framework.util.TreePathCacher` to `javacutils`
- extend TreePathCacher to handle `visitTry` in a similar fashion as the `BaseTypeVisitor.visitTry`
- use a TreePathCacher instead of `Trees.getPath` with an appropriate set of generated Trees.
- in particular, we can use a TreePathCacher from the CFGBuilder - maybe we can even share it between CFG builds.

@bitterfox does this sound right to you? Do you think you will have time to look into fixing this?

**Comment by smillst** [comment]
Make sure to add the test in #1586.

**Comment by smillst** [comment]
Fixed via 7c7391e72a14f1c5dd1870fbccbe4ed956bc3dbc
